### PR TITLE
Analytics

### PR DESCRIPTION
**Pull request by matzew**
updated version of https://github.com/aerogear/aerogear-unifiedpush-server/pull/540

### CODE REVIEW

File: model/api/src/main/java/org/jboss/aerogear/unifiedpush/dao/PushMessageInformationDao.java [hunk 1]
```diff
@@ -114,4 +114,7 @@
     List<String> findVariantIDsWithWarnings();
     List<PushMessageInformation> findLatestActivity(int maxResults);
     long getNumberOfPushMessagesForApplications();
+
+    PushMessageInformation getPushMessageInformation(String id);
```
Comment (matzew) [inline model/api/src/main/java/org/jboss/aerogear/unifiedpush/dao/PushMessageInformationDao.java]: @sebastienblanc what is the ID here, the PK ?

File: model/jpa/src/main/java/org/jboss/aerogear/unifiedpush/jpa/dao/impl/JPAPushMessageInformationDao.java [hunk 2]
```diff
@@ -141,6 +141,13 @@ public long getNumberOfPushMessagesForApplications() {
                 "IN (select p.pushApplicationID from PushApplication p)", Long.class).getSingleResult();
     }
 
+    @Override
+    public PushMessageInformation getPushMessageInformation(String id) {
+        return getSingleResultForQuery(createQuery(
+                "select pmi from PushMessageInformation pmi where pmi.id = :id")
+                .setParameter("id", id));
+
```
Comment (matzew) [inline model/jpa/src/main/java/org/jboss/aerogear/unifiedpush/jpa/dao/impl/JPAPushMessageInformationDao.java]: @sebastienblanc if that is the PK, why not using `em.find(PushMessageInformation.class, id);` ?

Comment (edewit) [inline model/jpa/src/main/java/org/jboss/aerogear/unifiedpush/jpa/dao/impl/JPAPushMessageInformationDao.java]: good find, but even better would be to replace it with [`find(id)`](https://github.com/aerogear/aerogear-unifiedpush-server/blob/d216bfa1a9a4fe9a9165874cec0ed79b79ab80d6/model/jpa/src/main/java/org/jboss/aerogear/unifiedpush/jpa/dao/impl/JPABaseDao.java#L60) so this entire method could be removed

Comment (matzew) [inline model/jpa/src/main/java/org/jboss/aerogear/unifiedpush/jpa/dao/impl/JPAPushMessageInformationDao.java]: ok, wil do

On Wed, May 6, 2015 at 3:06 PM, Erik Jan de Wit <notifications@github.com>
wrote:

> In
> model/jpa/src/main/java/org/jboss/aerogear/unifiedpush/jpa/dao/impl/JPAPushMessageInformationDao.java
> <https://github.com/aerogear/aerogear-unifiedpush-server/pull/543#discussion_r29755713>
> :
>
> > @@ -141,6 +141,13 @@ public long getNumberOfPushMessagesForApplications() {
> >                  "IN (select p.pushApplicationID from PushApplication p)", Long.class).getSingleResult();
> >      }
> >
> > +    @Override
> > +    public PushMessageInformation getPushMessageInformation(String id) {
> > +        return getSingleResultForQuery(createQuery(
> > +                "select pmi from PushMessageInformation pmi where pmi.id = :id")
> > +                .setParameter("id", id));
> > +
>
> good find, but even better would be to replace it with find(id)
> <https://github.com/aerogear/aerogear-unifiedpush-server/blob/d216bfa1a9a4fe9a9165874cec0ed79b79ab80d6/model/jpa/src/main/java/org/jboss/aerogear/unifiedpush/jpa/dao/impl/JPABaseDao.java#L60>
> so this entire method could be removed
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/aerogear/aerogear-unifiedpush-server/pull/543/files#r29755713>
> .
>



-- 
Matthias Wessendorf

blog: http://matthiaswessendorf.wordpress.com/
sessions: http://www.slideshare.net/mwessendorf
twitter: http://twitter.com/mwessendorf

**Comment by matzew** [comment]
@corinnekrych @edewit @danielpassos w/ https://github.com/matzew/aerogear-unifiedpush-server/commit/86134ea31f563f662441aabd4cef3107acba6594 we have a PUT endpoint for storing the "app opened due to push" ack.

A little CURL I did is:

```
curl -u "super:secret" -v -H "Accept: application/json" -H "Content-type: application/json" -X PUT http://127.0.0.1:8080/ag-push/rest/registry/device/pushMessage/eac8c56c-209e-4f20-95e4-c59404947bf6 
```

I have the endpoint under /registry/device for a few reasons:
* it's data sent from the actual device
* and no need to add another http basic exception

**Comment by matzew** [comment]
@secondsun ^ 

Now that we have a separate endpoint for the metrics - do you think we need a `PUT` for "updating" the registration metadata? or should the mobile clients be executing against our `POST` registration API for both: initial registration and continued/updated registration ?

**Comment by secondsun** [comment]
@matzew That kind of depends ;)

Doing a bit of research (IE bothering you in IRC) on Android and SimplePush the device installation id used by the server is provided by the client.  On Android we repeatedly call .register (because of a laundry list of technical reasons).  This invokes a POST request every time.  Because we know the ID and because multiple calls to register always leave the server in the same state we should probably be using PUT for both registration AND updating.

This is mostly a round about way of saying I think we should not create a second method for android setting updates and possibly, in the future, switch to using PUT instead of POST.

**Comment by matzew** [comment]
@secondsun given the amount of tickets we have for 1.1.0, I'd vote to not do that change for 1.1.0

I see your point, but being pragmatic it does not hurt to do POST for both: initial registration and updating registration data. 

I have filed a future JIRA: https://issues.jboss.org/browse/AGPUSH-1399

**Comment by secondsun** [comment]
@matzew  +1 for pragmatism

**Comment by cvasilak** [comment]
I am +1 for the separate endpoint for the metric analytic update. 
regarding the POST/PUT reg update can be looked later on I guess

:+1:

**Comment by danielpassos** [comment]
:+1:  for the separate endpoint

**Pull request merged**